### ISSUE TITLE
[JENKINS-42785] Parameterized Pipeline Projects have their Build Triggers unset after each build

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/pipeline-model-definition-plugin#1023]
<p>At least as of a few updates ago (I swear this used to work) I've noticed all the Pipeline projects I've set up only trigger once, and then I have to go in and set the build triggers again because all the triggers have become unset.</p>

<p>This of course makes the pipeline projects I've set up fairly unhelpful, since this thus requires manual intervention for every single run.</p>

<p>I've narrowed this down to parameters. If I have a "Pipeline script" or specify a Jenkinsfile with "Pipeline script from SCM", any build (whether triggered by one of the Build Triggers or manually) results in the Build Triggers becoming unset.</p>

<p>Specifically, the following settings, if set before the build is run, become uncheckmarked:</p>
<ul>
	<li>Build after other projects are built</li>
	<li>Build periodically</li>
	<li>GitHub hook trigger for GITScm polling</li>
	<li>Poll SCM</li>
</ul>

<p>Interestingly, although they are also under the Build Triggers heading, "Quiet period" and "Trigger builds remotely (e.g., from scripts)" remain set.</p>

<p>Here is an example Pipeline script that reproduces this problem for me:</p>

<pre>
pipeline {
    agent any
    parameters {
       string(name: <span class="code-quote">'requested'</span>, defaultValue: <span class="code-quote">'all'</span>, description: <span class="code-quote">'Which stages should be run?'</span>)
    }
    stages {
       stage(<span class="code-quote">'stage1'</span>) {
          steps {
             echo <span class="code-quote">'Running stage 1'</span>
          }
       }
    }
 }</pre>
<p> </p>

<p>Conversely, the following doesn't lose the scheduling settings:</p>

<pre>
pipeline {
    agent any
    stages {
       stage(<span class="code-quote">'stage1'</span>) {
          steps {
             echo <span class="code-quote">'Running stage 1'</span>
          }
       }
    }
 }</pre>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/keithzg.jpg" title="keithzg's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=keithzg">keithzg</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42785" target="_blank">Parameterized Pipeline Projects have their Build Triggers unset after each build</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer">abayer</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: pipeline-model-definition-plugin
<li><b>label(s)</b>: parameterized
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2017-07-06T12:12:58+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>At least as of a few updates ago (I swear this used to work) I've noticed all the Pipeline projects I've set up only trigger once, and then I have to go in and set the build triggers again because all the triggers have become unset.</p>

<p>This of course makes the pipeline projects I've set up fairly unhelpful, since this thus requires manual intervention for every single run.</p>

<p>I've narrowed this down to parameters. If I have a "Pipeline script" or specify a Jenkinsfile with "Pipeline script from SCM", any build (whether triggered by one of the Build Triggers or manually) results in the Build Triggers becoming unset.</p>

<p>Specifically, the following settings, if set before the build is run, become uncheckmarked:</p>
<ul>
	<li>Build after other projects are built</li>
	<li>Build periodically</li>
	<li>GitHub hook trigger for GITScm polling</li>
	<li>Poll SCM</li>
</ul>

<p>Interestingly, although they are also under the Build Triggers heading, "Quiet period" and "Trigger builds remotely (e.g., from scripts)" remain set.</p>

<p>Here is an example Pipeline script that reproduces this problem for me:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
    agent any
    parameters {
       string(name: <span class="code-quote">'requested'</span>, defaultValue: <span class="code-quote">'all'</span>, description: <span class="code-quote">'Which stages should be run?'</span>)
    }
    stages {
       stage(<span class="code-quote">'stage1'</span>) {
          steps {
             echo <span class="code-quote">'Running stage 1'</span>
          }
       }
    }
 }</pre>
</div></div>
<p> </p>

<p>Conversely, the following doesn't lose the scheduling settings:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
    agent any
    stages {
       stage(<span class="code-quote">'stage1'</span>) {
          steps {
             echo <span class="code-quote">'Running stage 1'</span>
          }
       }
    }
 }</pre>
</div></div></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Windows, Jenkins 2.50</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/orrc.svg" title="orrc's avatar" /> orrc:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/42785?focusedId=291911&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291911">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42785?focusedId=291911&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291911">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this specific to Declarative Pipelines, i.e. in the style you've written above, starting with <tt>pipeline {</tt>?</p>

<p>Just wondering whether this should be reassigned to the pipeline-model-definition component.</p>
</pre>
</details>

</ul>
</details>

<p>Is this specific to Declarative Pipelines, i.e. in the style you've written above, starting with <tt>pipeline {</tt>?</p>

<p>Just wondering whether this should be reassigned to the pipeline-model-definition component.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/keithzg.jpg" title="keithzg's avatar" /> keithzg:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/42785?focusedId=292661&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292661">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42785?focusedId=292661&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292661">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Apologies for the delayed reply, it's been unexpectedly busy at work. Unfortunately, I don't actually know how to write a <em>non</em>-declarative pipeline script, certainly not one that's parameterized, and I've had a hard time trying to find or work out such a script.</p>

<p>Edit: Looking at <a href="https://jenkins.io/doc/book/pipeline/syntax/#parameters" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/doc/book/pipeline/syntax/#parameters</a> it claims </p>
<div class='table-wrap'>
<table class='confluenceTable'><tbody>
<tr>
<td class='confluenceTd'>Only once, inside the <tt>pipeline</tt> block.</td>
</tr>
</tbody></table>
</div>


<p>which would seem to preclude using it outside of `pipeline {`? If not, certainly I can't discern any more, alas.</p>
</pre>
</details>

</ul>
</details>

<p>Apologies for the delayed reply, it's been unexpectedly busy at work. Unfortunately, I don't actually know how to write a <em>non</em>-declarative pipeline script, certainly not one that's parameterized, and I've had a hard time trying to find or work out such a script.</p>

<p>Edit: Looking at <a href="https://jenkins.io/doc/book/pipeline/syntax/#parameters" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/doc/book/pipeline/syntax/#parameters</a> it claims </p>
<div class='table-wrap'>
<table class='confluenceTable'><tbody>
<tr>
<td class='confluenceTd'>Only once, inside the <tt>pipeline</tt> block.</td>
</tr>
</tbody></table>
</div>


<p>which would seem to preclude using it outside of `pipeline {`? If not, certainly I can't discern any more, alas.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/42785?focusedId=293491&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-293491">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42785?focusedId=293491&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-293491">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>So...this brings up a good question! The root cause here is that job parameters, triggers and other misc job properties are all set in Pipeline (including Declarative, albeit behind the scenes) with the <tt>properties</tt> step...which overwrites whatever was previously in the job properties/triggers/parameters. So if you have <tt>parameters</tt> in your Declarative Pipeline but not <tt>triggers</tt>, your triggers are always going to get wiped out. The quiet period and remote trigger bit are technically not part of the <tt>triggers</tt> which is why they stay set. Is weird, I know.</p>

<p>So a workaround for this for you would be to use this as your <tt>Jenkinsfile</tt>:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
    agent any
    parameters {
       string(name: <span class="code-quote">'requested'</span>, defaultValue: <span class="code-quote">'all'</span>, description: <span class="code-quote">'Which stages should be run?'</span>)
    }
    triggers {
        cron(<span class="code-quote">'@daily'</span>)
        pollSCM(<span class="code-quote">'@hourly'</span>)
        githubPush()
        upstream(threshold: hudson.model.Result.SUCCESS, upstreamProjects: <span class="code-quote">"some-project,some-other-project"</span>)
    }
    stages {
       stage(<span class="code-quote">'stage1'</span>) {
          steps {
             echo <span class="code-quote">'Running stage 1'</span>
          }
       }
    }
 }</pre>
</div></div>
<p>I'm not 100% sure <tt>upstream</tt> will work right - it <b>should</b> but I honestly can't remember for sure. But the other three certainly will.</p>

<p>The question this all raises is whether Declarative should allow the UI-configured job properties/parameters/triggers to persist if they're not explicitly overridden. I'm not sure yet, which is why I'm leaving this open. =)</p>
</pre>
</details>

</ul>
</details>

<p>So...this brings up a good question! The root cause here is that job parameters, triggers and other misc job properties are all set in Pipeline (including Declarative, albeit behind the scenes) with the <tt>properties</tt> step...which overwrites whatever was previously in the job properties/triggers/parameters. So if you have <tt>parameters</tt> in your Declarative Pipeline but not <tt>triggers</tt>, your triggers are always going to get wiped out. The quiet period and remote trigger bit are technically not part of the <tt>triggers</tt> which is why they stay set. Is weird, I know.</p>

<p>So a workaround for this for you would be to use this as your <tt>Jenkinsfile</tt>:</p>

<pre>
pipeline {
    agent any
    parameters {
       string(name: <span class="code-quote">'requested'</span>, defaultValue: <span class="code-quote">'all'</span>, description: <span class="code-quote">'Which stages should be run?'</span>)
    }
    triggers {
cron(<span class="code-quote">'@​daily'</span>)
pollSCM(<span class="code-quote">'@​hourly'</span>)
githubPush()
upstream(threshold: hudson.model.Result.SUCCESS, upstreamProjects: <span class="code-quote">"some-project,some-other-project"</span>)
    }
    stages {
       stage(<span class="code-quote">'stage1'</span>) {
          steps {
             echo <span class="code-quote">'Running stage 1'</span>
          }
       }
    }
 }</pre>
<p>I'm not 100% sure <tt>upstream</tt> will work right - it <b>should</b> but I honestly can't remember for sure. But the other three certainly will.</p>

<p>The question this all raises is whether Declarative should allow the UI-configured job properties/parameters/triggers to persist if they're not explicitly overridden. I'm not sure yet, which is why I'm leaving this open. =)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/keithzg.jpg" title="keithzg's avatar" /> keithzg:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/42785?focusedId=293503&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-293503">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42785?focusedId=293503&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-293503">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Fair enough! Well for my part at least, I can confirm adding a triggers{} section solved the problem, so many thanks <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Fair enough! Well for my part at least, I can confirm adding a triggers{} section solved the problem, so many thanks <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/42785?focusedId=294175&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294175">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42785?focusedId=294175&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294175">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Ok, been thinking on this and I think we have to go with overwriting properties/triggers/parameters if they're not explicitly set in the Jenkinsfile. There's no way I can figure out to tell the difference between a property/trigger/parameter defined in the UI and one defined in the previous version of the Jenkinsfile but not in the current one, and we should definitely be removing anything in the latter scenario.</p>
</pre>
</details>

</ul>
</details>

<p>Ok, been thinking on this and I think we have to go with overwriting properties/triggers/parameters if they're not explicitly set in the Jenkinsfile. There's no way I can figure out to tell the difference between a property/trigger/parameter defined in the UI and one defined in the previous version of the Jenkinsfile but not in the current one, and we should definitely be removing anything in the latter scenario.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/42785?focusedId=294543&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294543">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42785?focusedId=294543&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294543">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It makes no sense to have both UI config and Jenkinsfile config is they fight and can't get along. </p>

<p>Ideally it would error or warn if the UI properties/triggers are "set" and Jenkinsfile is also "set". Or else remove it from the UI. Or the Jenkinsfile. as it is - not very nice. </p>
</pre>
</details>

</ul>
</details>

<p>It makes no sense to have both UI config and Jenkinsfile config is they fight and can't get along. </p>

<p>Ideally it would error or warn if the UI properties/triggers are "set" and Jenkinsfile is also "set". Or else remove it from the UI. Or the Jenkinsfile. as it is - not very nice. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bitwiseman.png" title="bitwiseman's avatar" /> bitwiseman:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/42785?focusedId=378388&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-378388">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42785?focusedId=378388&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-378388">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Bulk closing resolved issues. </p>
</pre>
</details>

</ul>
</details>

<p>Bulk closing resolved issues. </p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-44621%22">JENKINS-44621</a>]</i>
